### ISSUE TITLE
Need ability to trigger k8s jobs from within charms

### ISSUE THREAD

**Issue by jayk-canonical** [canonical/operator#292]
When deploying applications to Kubernetes, we need to be able to handle application lifecycle events, for example, database migrations, or other maintenance activities. This often requires the execution of code or containers to perform certain actions that are specific to the type of change occurring.

There does not appear to be a way to do this from within the operator framework.  We need this.

We discovered this issue during our attempt to implement a discourse charm.

The details are as follows: 

When performing certain actions, such as an update to the version of discourse, adding / removing plugins (changing the image) or certain other config changes, a migration step is necessary to apply the needed changes to the database.

This process should be run one time by a single unit when the change requires it.  As we have config_changed hooks, that would be an ideal place to trigger the job.  Unfortunately, as the operator framework does not have this ability, we could not do that.  

If we had a way to trigger a job and then be notified of it's completion, we could implement a process in the charm that triggers the required migration and once completed deploys the updated code or pod spec.

**TL;DR:**

_What I did:_
Attempted to create a charm that can handle database migrations upon configuration update.

_What I expected to happen:_
I expected that I could trigger an upgrade process upon a config change.

_What happened instead:_ 
I was unable to trigger the needed upgrade process.

_What else I tried:_
I attempted to work around this by implementing the migration during pod start, as well as in an initContainer, which partially works, but when the application has a higher scale than 1, as each pod runs the same startup / initContainer the upgrade/migrate processes collided and failed.

_Proposed solution:_ 
Provide a mechanism that allows a charm author to define and trigger a k8s job from within the charm and then receive notification that said job has completed. Optionally provide a way to declaratively define jobs to trigger on specific charm events (config_changed, etc.) 

Please let me know if you need any further information on this issue.

**Comment by jameinel** [comment]
Can you clarify what you mean by "trigger a job". Is this trying to kubectl exec from the operator pod, is this talking to the K8s API to be able to run a "Job" like: https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

As I understand it, a K8s "Job" is run as an independent pod, which when it completes the pod is destroyed.


From our earlier conversation, I think what you want is:

- The ability to trigger a script that exists inside the application pod to be run in response to a config-changed event
- The ability to control who runs that script and why (so that you can ask that a single pod runs the script, and not that all pods run the script.
- Right now you could do that via a Juju action (encapsulate the code you want to run inside a script, and trigger it to run on a particular unit). However, you cannot trigger actions from a charm, the only way to ask for an action to run is from the juju CLI.
- If this was a normal IAAS charm, then you'd just run the script. However, in K8s, the operator doesn't run on the same machine as the application, so you need a way to poke at the remote pod.


Alternatively, you might just be wanting the K8s api exposed inside the operator, such that you can "kubectl exec" a script from outside of that pod. Something like:
https://kubectl.docs.kubernetes.io/pages/container_debugging/executing_a_command_in_a_container.html
or maybe:
https://www.openshift.com/blog/executing-commands-in-pods-using-k8s-api

**Comment by jayk-canonical** [comment]
Hello!

Fundamentally we need to be able to run code in the context of the application container (or a separately provided container) in response to charm-level events.

In terms of how we accomplish this - I am open to what options might be available. In my original issue I was referring to a k8s job ( https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/ ) as that would be as close to the official k8s mechanism for this kind of workload. I think staying close to the k8s official way of doing this kind of thing will make it easier to control / port applications already built for k8s and would give us a better 'story' regarding using operator charms to control k8s.

That said, having the ability to trigger a 'kubectl exec' would be useful in it's own right and could be used to solve this problem.

**Comment by relaxdiego** [comment]
There is no need to use kubectl if you want to define and trigger a job. All pods have access to the API server directly via the token mounted in every pod. Provided that the pod has the necessary RBAC, you should be able to interact with the API Server as you wish. See this example:

https://github.com/relaxdiego/charm-k8s-prometheus/blob/1ca74541fbc32e083b45713c3d87c62b397e76d1/src/adapters/k8s.py#L38-L56

**Comment by jameinel** [comment]
We opened this discussion in discourse: https://discourse.juju.is/t/coordinating-actions-for-a-k8s-operators/3161

The discussion there is that 'juju-run' should be available to Charms, but probably can't be run in the middle of a hook (you might be able to unset JUJU_CONTEXT_ID, but that might run afoul of trying to take out a hook/machine lock while you're already running one).

The claim is also that the RBAC on K8s api should already allow you to run a pod as a K8s Job if that is the preferred path forward.

**Comment by rwcarlsen** [comment]
Now that we've settled primarily on sidecar k8s charms moving forward, I'm closing this as resolved - we have pebble exec and such to do this in that world.
